### PR TITLE
Codechange: Don't use a dense matrix to store LinkGraph edges.

### PR DESCRIPTION
**Pull request by michicc**
## Motivation / Problem

The LinkGraph uses a dense matrix to store information about links (edges) between stations (nodes).
This matrix can be potentially big and contains mostly unused entries since it is not used to store distances anymore (see #6041).
The savegame is already only storing the used entries of the matrix.

Additionally, the source and destination of each edge are purely determined positionally, which necessitates wrapper classes around nodes and edges and custom iterator implementations to walk the edges, just so the callers aren't exposed to this.
The `SmallMatrix` container used to store the edges is also not C++-aware, which is worked around by simulating constructors.

## Description

This PR moves the edges to be stored into a simple sorted `std::vector` in each node.
Each edge also stores its destination, so there's no need for any custom wrappers and iterators anymore.
The annotations of the LinkGraphJob are given reference to the source elements, removing the position dependence there, too.

As the linkgraph was the only user of `SmallMatrix`, it can go, too.

## Limitations

Not tested for performance.
Sparsely used dense matrices are usually not very cache-friendly, so not having them is probably good. Also, resizing the matrix is expensive.

Minimally checked for correctness as I am not an avid cdist player and don't have a large sample set of reference savegames at hand.

## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR touches english.txt or translations? Check the [guidelines](https://github.com/OpenTTD/OpenTTD/blob/master/docs/eints.md)
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, gs_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

### CODE REVIEW

**Comment by LordAro** [comment]
>Not tested for performance.

Would be very useful to get some numbers on this

File: src/saveload/saveload.h [hunk 1]
```diff
@@ -344,6 +344,8 @@ enum SaveLoadVersion : uint16 {
 	SLV_LAST_LOADING_TICK,                  ///< 301  PR#9693 Store tick of last loading for vehicles.
 	SLV_MULTITRACK_LEVEL_CROSSINGS,         ///< 302  PR#9931 v13.0  Multi-track level crossings.
 
+	SLV_LINKGRAPH_EDGES,                    ///< 303  PR#xxxxx Explicitly store link graph edges destination.
```
Comment (LordAro) [inline src/saveload/saveload.h]: you can do this now :)

**Comment by michicc** [comment]
> Would be very useful to get some numbers on this

Anybody got some good test savegames for this? I don't play much with cargodist (well, more like at all), so my collection is severely lacking for something with noticeable linkgraph CPU load.

**Comment by JGRennison** [comment]
I've pinched some ideas from here into my branch, and in doing so have added some extra things which might be worth considering here.

The number of EdgeAnnotation instances per NodeAllocation is relatively small, and does not change after initialisation.
To reduce the allocation overheads of std::vector instances per NodeAllocation, you could allocate all the EdgeAnnotation instances in a single std::vector or other contiguous block, and use std::span instances within NodeAllocation. This would probably get you better memory locality.
You can do the same thing for DemandAnnotation instances once the demand allocation phase is done, but this is a bit more fiddly.

In EdgeAnnotation, the `const LinkGraph::BaseEdge &base` reference could be removed outright.
It is (in most cases) 8 bytes, and only things which are actually needed through it are the capacity and the distance annotation (which can be precomputed), which also total to 8 bytes. This also would probably get you better memory locality.

**Review by LordAro — APPROVED** [review]
@JGRennison PRing your additional changes would be appreciated too

**Pull request merged**